libraries/tock-register-interface/README.md

```diff
@@ -3,130 +3,107 @@
-for each register an offset, a field name, and a type. Registers must be
-declared in increasing order of offsets and contiguously. Gaps when defining the
-registers must be explicitly annotated with an offset and gap identifier (by
-convention using a field named `_reservedN`), but without a type. The macro will
-then automatically take care of calculating the gap size and inserting a
-suitable filler struct. The end of the struct is marked with its size and the
-`@END` keyword, effectively pointing to the offset immediately past the list of
-registers.
+## Defining peripherals
 
-```rust
-use tock_registers::registers::{ReadOnly, ReadWrite, WriteOnly};
+An MMIO peripheral is defined using the `peripheral!` macro:
 
-register_structs! {
+```rust
+tock_registers::peripheral! {
     Registers {
-        // Control register: read-write
-        // The 'Control' parameter constrains this register to only use fields from
-        // a certain group (defined below in the bitfields section).
-        (0x000 => cr: ReadWrite<u8, Control::Register>),
-
-        // Status register: read-only
-        (0x001 => s: ReadOnly<u8, Status::Register>),
-
-        // Registers can be bytes, halfwords, or words:
-        // Note that the second type parameter can be omitted, meaning that there
-        // are no bitfields defined for these registers.
-        (0x002 => byte0: ReadWrite<u8>),
-        (0x003 => byte1: ReadWrite<u8>),
-        (0x004 => short: ReadWrite<u16>),
+        // Control register: read-write. `u8` specifies the data type of the
+        // memory-mapped register, and the `Control` parameter specifies how
+        // bitfields within the register should be accessed.
+        0x000 => cr: u8(Control::Register) { Read, Write },
+
+        // Status register: read-only. Registers can have documentation
+        // comments, like:
+        /// Status register
+        0x001 => s: u8(Status::Register) { Read },
+
+        // Registers can have different sizes. The bitfield parameter is
+        // optional:
+        0x002 => byte0: u8 { Read, Write },
+        0x003 => byte1: u8 { Read, Write },
+        0x004 => short: u16 { Read, Write },
 
         // Empty space between registers must be marked with a padding field,
-        // declared as follows. The length of this padding is automatically
-        // computed by the macro.
-        (0x006 => _reserved),
-        (0x008 => word: ReadWrite<u32>),
-
-        // The type for a register can be anything. Conveniently, you can use an
-        // array when there are a bunch of similar registers.
-        (0x00C => array: [ReadWrite<u32>; 4]),
-        (0x01C => ... ),
+        // specified by naming the field _ and omitting the operations list:
+        0x006 => _: u16,
 
-        // Etc.
+        // Padding's length may also be inferred by the macro by specifying _ as
+        // the data type:
+        0x008 => _: _,
```

Comment:
Does this include arbitrarily long gaps? If the gap is 100 bytes, will the system automatically make an array of padding?

Do you need to have `0x008 => _: _,` at all? What would happen if you did:
```
0x006 => byte0: u8 {Read, Write},
// gap here
0x00C => byte1: u8 {Read, Write},
```
Presumably that would be a compile-time failure?

But this code would work?
```
0x006 => byte0: u8 {Read, Write},
// HUGE gap here
0x007 => _: _,
0x100 => byte1: u8 {Read, Write},
```

Reply:
Oops, I was originally planning to have users specify the size of the padding (syntax A):

```rust
0x0 => ctrl: u8 { Write },
0x1 => _: u8, // One byte of padding
0x2 => status: u16 { Read },
```

however, when I was working on the README, I realized the current design doesn't require the user to specify the size:

```rust
(0x0 => ctrl: WriteOnly<u8>),
(0x1 => _reserved),
(0x2 => status: ReadOnly<u16>),
```

I want to migrate as many uses away from `register_structs!` as possible (because it is unsound), so I want to make the new syntax no less powerful than `register_structs!` (apart from being MMIO-only, which was a necessary concession to keep the complexity reasonable). I originally tweaked the syntax to optionally allow the size to be inferred, which would've looked like (syntax B):

```rust
0x0 => ctrl: u8 { Write },
0x1 => _: _, // Inferred padding (length will be 0x2 - 0x1)
0x2 => status: u16 { Read },
```

However I ultimately decided to simplify it and remove the length entirely, making the current syntax (syntax C):

```rust
0x0 => ctrl: u8 { Write },
0x1 => _, // Inferred padding (length will be 0x2 - 0x1)
0x2 => status: u16 { Read },
```

However, I apparently missed a couple cases of syntax B in README when I made that change. I've updated the README and `peripheral!`'s docs to show syntax C, which is my current design. Sorry about the confusion, hopefully that clears some things up!

Reply:
I like syntax C, with future expansion to allow `0x1 => _reserved_with_this_name: [u32; 4],`